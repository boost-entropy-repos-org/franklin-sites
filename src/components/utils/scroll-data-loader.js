const scrollOffsetPercentage = 10;
const scrollOffsetFactor = 1 + scrollOffsetPercentage / 100;

export const isNotScrollable = (node) => {
  const { scrollHeight, offsetHeight } = node.current;
  return scrollHeight <= offsetHeight * scrollOffsetFactor;
};

export const isBottom = (node) => {
  const { scrollHeight, scrollTop, offsetHeight } = node.current;
  return scrollHeight - Math.ceil(scrollTop) <= offsetHeight * scrollOffsetFactor;
};

export const checkLoadMoreItems = (loading, hasMoreData, ref, setLoadMoreItems) => {
  if (!loading && hasMoreData && isBottom(ref)) {
    setLoadMoreItems(true);
  }
};

export const checkOnDataLoad = (
  hasMoreData,
  setLoading,
  onLoadMoreItems,
  setLoadMoreItems,
  ref,
) => {
  if (hasMoreData && isNotScrollable(ref)) {
    setLoading(true);
    onLoadMoreItems();
  } else {
    setLoading(false);
    setLoadMoreItems(false);
  }
};

export const checkOnLoadMoreItems = (loadMoreItems, setLoading, onLoadMoreItems) => {
  if (loadMoreItems) {
    setLoading(true);
    onLoadMoreItems();
  }
};
