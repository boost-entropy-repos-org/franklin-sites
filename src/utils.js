export function getLastIndexOfSubstringIgnoreCase(string, substring) {
  return string.toLowerCase().lastIndexOf(substring.toLowerCase());
}

export const getFlattenedPaths = (currentItems, value, path = []) => {
  let flattened = [];
  currentItems.forEach((node) => {
    const { items, ...thisNode } = node;
    const nodePath = [...path, thisNode];
    if (items) {
      const result = getFlattenedPaths(items, value, nodePath);
      if (result.length) {
        flattened = [...flattened, ...result];
      }
    } else if (!value || thisNode.value === value) {
      flattened = [...flattened, nodePath];
    }
  });
  return flattened;
};

function getSquashedPath(items, sep = ' / ') {
  return {
    value: items[items.length - 1].value,
    pathLabel: items.map(item => item.label).join(sep),
    itemLabel: items[items.length - 1].label,
    items,

  };
}

export function getSquashedArrayOfPaths(paths) {
  // console.log(paths);
  return paths.map(items => getSquashedPath(items));
}
