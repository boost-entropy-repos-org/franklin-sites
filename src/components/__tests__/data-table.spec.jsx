import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import DataTable from '../data-table';
import { fillArray } from '../../utils';

describe('DataTable', () => {
  const idKey = 'id';
  const onSelect = jest.fn();
  const onHeaderClick = jest.fn();
  const selectable = true;
  const selected = {};
  const hasMoreData = true;
  const columns = [
    {
      label: 'Column 1',
      name: 'content1',
      render: row => row.content1,
      sortable: true,
      sorted: 'ascend',
    },
    {
      label: 'Column 2',
      name: 'content2',
      render: row => row.content2,
    },
    {
      label: 'Column 3',
      name: 'content3',
      render: row => row.content3,
      sortable: true,
    },
  ];
  const data = fillArray(10, (element, index) => ({
    [idKey]: `id-${index}`,
    content1: 'foo',
    content2: 'bar',
    content3: 'baz',
  }));

  let onLoadMoreItems;
  let props;
  beforeEach(() => {
    onLoadMoreItems = jest.fn();
    props = {
      onLoadMoreItems,
      hasMoreData,
      data,
      idKey,
      columns,
      onSelect,
      selected,
      onHeaderClick,
      selectable,
    };
  });

  test('should render', () => {
    const { asFragment } = render(<DataTable {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('should request more data', () => {
    const { container } = render(<DataTable {...props} />);
    const scrollContainer = container.firstChild;
    fireEvent.scroll(scrollContainer, { target: { scrollY: 1000 } });
    expect(onLoadMoreItems).toHaveBeenCalled();
  });

  test('should not request more data when hasMoreData is false', () => {
    const { container } = render(<DataTable {...props} hasMoreData={false} />);
    const scrollContainer = container.firstChild;
    fireEvent.scroll(scrollContainer, { target: { scrollY: 1000 } });
    expect(onLoadMoreItems).toHaveBeenCalledTimes(0);
  });

  test('should fire onSelect when checkbox is clicked', () => {
    const { container } = render(<DataTable {...props} />);
    const checkbox = container.querySelector('tbody tr td input');
    fireEvent.click(checkbox);
    expect(onSelect).toHaveBeenCalled();
  });

  test('should fire onHeaderClick when header is clicked', () => {
    const { container } = render(<DataTable {...props} />);
    const sortableHeader = container.querySelector('.data-table__table__header__row__cell--ascend');
    fireEvent.click(sortableHeader);
    expect(onHeaderClick).toHaveBeenCalled();
  });
});
