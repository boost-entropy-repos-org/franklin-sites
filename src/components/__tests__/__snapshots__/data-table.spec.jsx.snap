// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTable should render 1`] = `
<DocumentFragment>
  <div
    class="data-loader__wrapper"
  >
    <div
      class="data-loader__scroll-container"
    >
      <table
        class="data-table   "
      >
        <thead
          class="data-table__header"
        >
          <tr
            class="data-table__row"
          >
            <th
              class="data-table__header-cell data-table__header-cell--checkbox"
            >
               
            </th>
            <th
              class="data-table__header-cell data-table__header-cell--ascend"
              style="width: auto;"
            >
              Column 1
            </th>
            <th
              class="data-table__header-cell "
              style="width: auto;"
            >
              Column 2
            </th>
            <th
              class="data-table__header-cell data-table__header-cell--sortable"
              style="width: auto;"
            >
              Column 3
            </th>
          </tr>
        </thead>
        <tbody
          class="data-table__table__body"
        >
          <tr>
            <td
              class="data-table__cell data-table__cell--selected"
            >
              <input
                checked=""
                type="checkbox"
              />
            </td>
            <td
              class="data-table__cell data-table__cell--selected "
            >
              foo
            </td>
            <td
              class="data-table__cell data-table__cell--selected "
            >
              bar
            </td>
            <td
              class="data-table__cell data-table__cell--selected "
            >
              baz
            </td>
          </tr>
          <tr>
            <td
              class="data-table__cell data-table__cell--odd "
            >
              <input
                type="checkbox"
              />
            </td>
            <td
              class="data-table__cell data-table__cell--odd  "
            >
              foo
            </td>
            <td
              class="data-table__cell data-table__cell--odd  "
            >
              bar
            </td>
            <td
              class="data-table__cell data-table__cell--odd  "
            >
              baz
            </td>
          </tr>
          <tr>
            <td
              class="data-table__cell "
            >
              <input
                type="checkbox"
              />
            </td>
            <td
              class="data-table__cell  "
            >
              foo
            </td>
            <td
              class="data-table__cell  "
            >
              bar
            </td>
            <td
              class="data-table__cell  "
            >
              baz
            </td>
          </tr>
          <tr>
            <td
              class="data-table__cell data-table__cell--odd "
            >
              <input
                type="checkbox"
              />
            </td>
            <td
              class="data-table__cell data-table__cell--odd  "
            >
              foo
            </td>
            <td
              class="data-table__cell data-table__cell--odd  "
            >
              bar
            </td>
            <td
              class="data-table__cell data-table__cell--odd  "
            >
              baz
            </td>
          </tr>
          <tr>
            <td
              class="data-table__cell "
            >
              <input
                type="checkbox"
              />
            </td>
            <td
              class="data-table__cell  "
            >
              foo
            </td>
            <td
              class="data-table__cell  "
            >
              bar
            </td>
            <td
              class="data-table__cell  "
            >
              baz
            </td>
          </tr>
          <tr>
            <td
              class="data-table__cell data-table__cell--odd "
            >
              <input
                type="checkbox"
              />
            </td>
            <td
              class="data-table__cell data-table__cell--odd  "
            >
              foo
            </td>
            <td
              class="data-table__cell data-table__cell--odd  "
            >
              bar
            </td>
            <td
              class="data-table__cell data-table__cell--odd  "
            >
              baz
            </td>
          </tr>
          <tr>
            <td
              class="data-table__cell "
            >
              <input
                type="checkbox"
              />
            </td>
            <td
              class="data-table__cell  "
            >
              foo
            </td>
            <td
              class="data-table__cell  "
            >
              bar
            </td>
            <td
              class="data-table__cell  "
            >
              baz
            </td>
          </tr>
          <tr>
            <td
              class="data-table__cell data-table__cell--odd "
            >
              <input
                type="checkbox"
              />
            </td>
            <td
              class="data-table__cell data-table__cell--odd  "
            >
              foo
            </td>
            <td
              class="data-table__cell data-table__cell--odd  "
            >
              bar
            </td>
            <td
              class="data-table__cell data-table__cell--odd  "
            >
              baz
            </td>
          </tr>
          <tr>
            <td
              class="data-table__cell "
            >
              <input
                type="checkbox"
              />
            </td>
            <td
              class="data-table__cell  "
            >
              foo
            </td>
            <td
              class="data-table__cell  "
            >
              bar
            </td>
            <td
              class="data-table__cell  "
            >
              baz
            </td>
          </tr>
          <tr>
            <td
              class="data-table__cell data-table__cell--odd "
            >
              <input
                type="checkbox"
              />
            </td>
            <td
              class="data-table__cell data-table__cell--odd  "
            >
              foo
            </td>
            <td
              class="data-table__cell data-table__cell--odd  "
            >
              bar
            </td>
            <td
              class="data-table__cell data-table__cell--odd  "
            >
              baz
            </td>
          </tr>
        </tbody>
      </table>
      <div
        class="data-loader__loading"
      >
        Loading...
      </div>
    </div>
  </div>
</DocumentFragment>
`;
